OCI version range support

scottrigby opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Make sure to read the Contributing Guide before submitting your PR: https://github.com/helm/helm/blob/main/CONTRIBUTING.md
2. If this PR closes another issue, add 'closes #<issue number>' somewhere in the PR summary. GitHub will automatically close that issue when this PR gets merged. Alternatively, adding 'refs #<issue number>' will not close the issue, but help provide the reviewer more context.-->

**What this PR does / why we need it**:

closes #9694 

**Special notes for your reviewer**:

- Thank you to @mattfarina and @jdolitsky for co-navigating this initially
- Collaborating on this PR branch of my fork with @sabre1041

**If applicable**:
- ~[ ] this PR contains documentation~
- [x] this PR contains unit tests
- [x] this PR has been tested for backwards compatibility

scottrigby committed:
hack in progress

Signed-off-by: Scott Rigby <scott@r6by.com>

scottrigby committed:
Temp use Andy's fork PR branch

Signed-off-by: Scott Rigby <scott@r6by.com>

scottrigby committed:
It appears we never got to this block below. Quick rec by Farina. Untested if necessary

Signed-off-by: Scott Rigby <scott@r6by.com>

scottrigby committed:
Revert "hack in progress"

This reverts commit c0be414e4b8d2928018504c010cb04b1b2450bf3.

Taking a different approach, but keep this work in git history for now until we
know the new approach works.

Signed-off-by: Scott Rigby <scott@r6by.com>

scottrigby committed:
Revert "It appears we never got to this block below. Quick rec by Farina. Untested if necessary"

This reverts commit f616a01808da3428c6191e5196f32ca72eb22254.

Removing for now until we know we need it.

Signed-off-by: Scott Rigby <scott@r6by.com>

sabre1041 committed:
Initial tag listing support

Signed-off-by: Andrew Block <andy.block@gmail.com>

scottrigby committed:
Add OCI tag verions to the Dependency object before Resolve. TODO: fix HTTP HTTPS error for local registries

Signed-off-by: Scott Rigby <scott@r6by.com>

scottrigby committed:
Update ORAS to v1.1.0-rc1

Now that https://github.com/oras-project/oras-go/pull/89 is merged and released

Signed-off-by: Scott Rigby <scott@r6by.com>

scottrigby committed:
Fix Trim to TrimPrefix

Signed-off-by: Scott Rigby <scott@r6by.com>

mattfarina commented on pkg/downloader/manager.go:
```diff
@@ -233,6 +233,26 @@ func (m *Manager) loadChartDir() (*chart.Chart, error) {
+	// However we can add a new backward compatible struct field to
+	// *chart.Dependency, so that we can pass these along to
+	// (*Resolver).resolve from here through reqs []*chart.Dependency.
+	for i, d := range req {
+		if registry.IsOCI(d.Repository) {
+			// TODO(scottrigby): fix HTTP error
+			// > Error: Get "https://localhost:5000/v2/helm-charts/tags/list": http: server gave HTTP response to HTTPS client
+			ref := strings.TrimPrefix(d.Repository, "oci://")
+			tags, err := m.RegistryClient.Tags(ref)
+			if err != nil {
+				return nil, err
+			}
+			req[i].OCITagVersions = tags
+		}
+	}
```
Why was this additional loop added here? This could be done in `internal/resolver/resolver.go` at the same place where `OCITagVersions` is used. Then the `Dependency` struct doesn't need to have any new property put on it. Localizing the work to the resolver would make the change simpler with fewer possible complications to consider.

scottrigby replied:
The problem is I don't see a backward compatible way to be able to get `RegistryClient` from within (*Resolver).Resolve, which is needed to call the API. However the calling function `(*Manager).resolve` has this already. So making the tag API call from there, then attaching the result to `Dependency` object was a solution for this that is backward compatible. Do you see another way?

mattfarina replied:
The resolver is an [`internal`](https://go.dev/doc/go1.4#internalpackages) package and not part of the public API. Take a look at its location. This is not a general purpose resolver meant to be imported and used by others. Its only access is through the `resolve` method here. It's meant just to support Helms specific use.

You already added a 3rd argument to `resolver.New` to pass in the registry client and you extended `Resolver` to hold on to it. Now you just need to use it within the `Resolve` method.

sabre1041 replied:
@mattfarina thats exactly the approach that I am working on today

scottrigby replied:
Thanks @mattfarina you're right – and yes that makes it easier! @sabre1041 & I talked about this approach last night. Yep I see Andy added that now. Much better

scottrigby replied:
In touch w Andy on this, so we will get it done tonight one way or another 💯

mattfarina commented on internal/experimental/registry/client.go:
```diff
@@ -511,3 +524,18 @@ func PushOptStrictMode(strictMode bool) PushOption {
 	}
 }
+
+// Tags lists all tags for a given repository
```
It might be useful to note that the response of tags isn't sorted.

scottrigby replied:
No longer true since https://github.com/helm/helm/pull/10527/commits/348942dfc79b193a08c700d0c69eebc6f5eef3b7. Yay. See https://github.com/helm/helm/pull/10527#discussion_r780831755

mattfarina commented on internal/resolver/resolver.go:
```diff
@@ -139,6 +139,20 @@ func (r *Resolver) Resolve(reqs []*chart.Dependency, repoNames map[string]string
+
+			// This works, fake it to test mocking the version
+			// tags := []string{"0.1.0", "0.1.1", "0.2.0"}
+			tags := d.OCITagVersions
```
The tags here aren't sorted. I didn't test it but reading through the oras tag code it doesn't appear to be. This list is turned into `ChartVersion` which are still unsorted by version. But, the version constraint handling expects it to be sorted. So, these tags need to be sorted and non-semver tags pulled out as part of the process. This will be especially important because there is work to support sigstore which uses tags.

You can find a sorting example at https://github.com/Masterminds/semver/blob/master/collection_test.go (tests as an example)

scottrigby replied:
Thanks this is helpful. Will do soon.

scottrigby replied:
Been collabing w Andy, who just pushed a fix for this - added directly to registry `(*Client).tags` 💯

mattfarina reviewed (changes requested):
Found some issues with the current setup. I added suggestions on how to simplify the structure and fix the issues.

sabre1041 committed:
Added registryClient to resolver

Signed-off-by: Andrew Block <andy.block@gmail.com>

sabre1041 committed:
Updated tag resolver logic

Signed-off-by: Andrew Block <andy.block@gmail.com>

sabre1041 committed:
Working oci code without providing versions

Signed-off-by: Andrew Block <andy.block@gmail.com>

sabre1041 committed:
Started work on 'helm pull'

Signed-off-by: Andrew Block <andy.block@gmail.com>

mattfarina committed:
Fixing issue where OCI handling early causes a bad message

Note, there is OCI handling later in the funtion that should
handle the situation instead.

Closes #10534

Signed-off-by: Matt Farina <matt.farina@suse.com>

scottrigby committed:
Implement reusable GetTagMatchingVersionOrConstraint

Largely borrowed from (IndexFile).Get. However there is not currently a nice
way to make this code also usable to the repo package, as IndexFile depends on
a list of index Entries containing a nexted version.

We could refactor this later to somehow use the same shared function, but for
now keeping separate.

Signed-off-by: Scott Rigby <scott@r6by.com>

scottrigby committed:
Fix import

Signed-off-by: Scott Rigby <scott@r6by.com>

scottrigby commented:
Merging this PR is on hold until the HTTP HTTPS auth issue is addressed. We are planning to connect with @deitch tomorrow to get direction on the best way to manage auth and http configuration on the registry client with those defined on an oras client. Slack convo [here](https://cloud-native.slack.com/archives/CJ1KHJM5Z/p1641842277021100).

Apart from that though, this pr is working fine with HTTPS and no auth. Both for chart OCI dependencies using semver range, and helm install semver logic. See the manual test output below testing with an example HTTPS publicly accessible repo (https://github.com/sigstore/helm-charts/pkgs/container/helm-charts%2Frekor/versions ). Looks correct to me. Feel free to test for yourself with different semver ranges ([see supported semver constraint definitions here](https://github.com/Masterminds/semver)).

```sh
$ make

# Ranges
$ ./bin/helm install --dry-run oci://ghcr.io/sigstore/helm-charts/rekor --generate-name --set server.ingress.enabled=false --version "<0.2.10" | grep helm.sh/chart | head -n1
    helm.sh/chart: rekor-0.2.

$ ./bin/helm install --dry-run oci://ghcr.io/sigstore/helm-charts/rekor --generate-name --set server.ingress.enabled=false --version ">0.2.9"| grep helm.sh/chart | head -n1
    helm.sh/chart: rekor-0.2.10

# PATCH wildcard
$ ./bin/helm install --dry-run oci://ghcr.io/sigstore/helm-charts/rekor --generate-name --set server.ingress.enabled=false --version "0.2.*" | grep helm.sh/chart | head -n1
    helm.sh/chart: rekor-0.2.10

# Missing tags
$ ./bin/helm install --dry-run oci://ghcr.io/sigstore/helm-charts/rekor --generate-name --set server.ingress.enabled=false --version "artifacthub.io" | grep helm.sh/chart | head -n1
Error: INSTALLATION FAILED: failed to download "oci://ghcr.io/sigstore/helm-charts/rekor" at version "artifacthub.io"

$ ./bin/helm install --dry-run oci://ghcr.io/sigstore/helm-charts/rekor --generate-name --set server.ingress.enabled=false --version "0.1.*" | grep helm.sh/chart | head -n1
Error: INSTALLATION FAILED: failed to download "oci://ghcr.io/sigstore/helm-charts/rekor" at version "0.1.*"

# No version passed
$ ./bin/helm install --dry-run oci://ghcr.io/sigstore/helm-charts/rekor --generate-name --set server.ingress.enabled=false | grep helm.sh/chart | head -n1
    helm.sh/chart: rekor-0.2.10
```

sabre1041 committed:
Enabled auth and support http registries for OCI

Signed-off-by: Andrew Block <andy.block@gmail.com>

sabre1041 committed:
Readded resolver OCI logic

Signed-off-by: Andrew Block <andy.block@gmail.com>

sabre1041 committed:
Fixed bad commit

Signed-off-by: Andrew Block <andy.block@gmail.com>

mattfarina commented on pkg/downloader/manager.go:
```diff
@@ -578,7 +578,8 @@ func (m *Manager) resolveRepoNames(deps []*chart.Dependency) (map[string]string,
 	for _, dd := range deps {
 		// Don't map the repository, we don't need to download chart from charts directory
-		if dd.Repository == "" {
+		// When OCI is used there is no Helm repository
+		if dd.Repository == "" || registry.IsOCI(dd.Repository) {
```
You might want to rebase off `main` or otherwise undo this change. Adding this back causes an error.

scottrigby replied:
Yes thank you, will do

scottrigby replied:
cherry-picked ee73a0263c02a6c53261fdaf2ad97542b18a844c from https://github.com/helm/helm/pull/10535 back into this branch 👍

mattfarina commented on pkg/action/pull.go:
```diff
@@ -92,13 +92,13 @@ func (p *Pull) Run(chartRef string) (string, error) {
-			getter.WithTagName(p.Version))
+			getter.WithRegistryClient(p.cfg.RegistryClient))
+
+		if p.Version != "" {
+			c.Options = append(c.Options,
+				getter.WithTagName(p.Version))
+		}
```
The version here isn't a version but can be a specific version or range. In either case, it's not a tag. So, using `WithTagName` don't convey what it does. As an external API to API users this is confusing.

Also, the other built-in getters don't turn a version range to a version in this part of the code. Doing it here makes it different from the others. I would pull the version part from here. See my other comments on where to put the version handling.

scottrigby replied:
done

mattfarina commented on pkg/getter/ocigetter.go:
```diff
@@ -50,10 +52,28 @@ func (g *OCIGetter) get(href string) (*bytes.Buffer, error) {
 
-	if version := g.opts.version; version != "" {
-		ref = fmt.Sprintf("%s:%s", ref, version)
+	// Retrieve list of repository tags
+	tags, err := client.Tags(ref)
+	if err != nil {
+		return nil, err
+	}
+	if len(tags) == 0 {
+		return nil, errors.Errorf("Unable to locate any tags in provided repository: %s", ref)
+	}
+
+	// Determine if version provided
+	// If empty, try to get the highest available tag
+	// If exact version, try to find it
+	// If semver constraint string, try to find a match
+	providedVersion := g.opts.version
+
+	tag, err := registry.GetTagMatchingVersionOrConstraint(tags, providedVersion)
+	if err != nil {
+		return nil, err
 	}
 
+	ref = fmt.Sprintf("%s:%s", ref, tag)
```
Sorting out the version for the other downloaders happens in another place. In `DownloadTo` a version or range is passed. The first thing `DownloadTo` does is call `c.ResolveChartVersion` which turns a version into a URI to call. This is then passed off to a getter which does the thing it was designed to do... get the thing.

Shouldn't this getter function using a similar code flow so the logic is in the same places?

scottrigby replied:
done

scottrigby committed:
Update oras-go to v1.1.0

Signed-off-by: Scott Rigby <scott@r6by.com>

scottrigby committed:
Move OCI tag semver range logic from OCIGetter to ChartDownloader

Signed-off-by: Scott Rigby <scott@r6by.com>

scottrigby committed:
Fix linting

Signed-off-by: Scott Rigby <scott@r6by.com>

sabre1041 committed:
Updates to chart downloader

Signed-off-by: Andrew Block <andy.block@gmail.com>

sabre1041 committed:
Handling name of OCI file

Signed-off-by: Andrew Block <andy.block@gmail.com>

mattfarina reviewed (changes requested):
:(

Resolving and pulling images with build metadata fails. For example, if I try to pull `0.3.0+build.1` and another one of `0.3.0` exists it will pull the one without the build metadata.

scottrigby committed:
Bring exact version check logic from IndexFile.Get into registry tag check

Signed-off-by: Scott Rigby <scott@r6by.com>

scottrigby committed:
Remove unneeded assignment

Signed-off-by: Scott Rigby <scott@r6by.com>

mattfarina commented on internal/experimental/registry/client.go:
```diff
@@ -539,3 +572,53 @@ func PushOptStrictMode(strictMode bool) PushOption {
+
+	var tagVersions []*semver.Version
+	for _, tag := range registryTags {
+		tagVersion, err := semver.StrictNewVersion(tag)
```
```
tagVersion, err := semver.StrictNewVersion(strings.ReplaceAll(tag, "_", "+"))
```

Need to make sure the _ has been turned back into a + before parsing the version

scottrigby replied:
Ah, thanks. I even left a note under "special notes" https://github.com/helm/helm/pull/10521 so we didn’t forget to make sure to add back to this PR. Guess what, I forgot 😛

scottrigby committed:
Change underscore (_) back to plus (+) for Helm

Signed-off-by: Scott Rigby <scott@r6by.com>

scottrigby commented:
Can not emphasize enough how much work went into this and how good the team collaboration was

# 👏 🥇 🎉
